File: app/src/main/java/com/work/criminalintent/CrimeLab.java
package com.work.criminalintent;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.work.criminalintent.database.CrimeBaseHelper;
import com.work.criminalintent.database.CrimeCursorWrapper;
import com.work.criminalintent.database.CrimeDbSchema;
import com.work.criminalintent.database.CrimeDbSchema.CrimeTable;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class CrimeLab {
    private static CrimeLab sCrimelab;
    private Context mContext;
    private SQLiteDatabase mDatabase;

    private CrimeLab(Context context) {
        mContext = context.getApplicationContext();
        mDatabase = new CrimeBaseHelper(mContext).getWritableDatabase();

    }

    public static CrimeLab get(Context context) {
        if (sCrimelab == null) {
            sCrimelab = new CrimeLab(context);
        }
        return sCrimelab;
    }

    public List<Crime> getCrimes() {
        List<Crime> crimes = new ArrayList<>();

        try (CrimeCursorWrapper cursor = queryCrimes(null, null)) {
            cursor.moveToFirst();
            while (!cursor.isAfterLast()) {
                crimes.add(cursor.getCrime());
                cursor.moveToNext();
            }
        }
        return crimes;
    }

    public Crime getCrime(UUID uuid) {
        try (CrimeCursorWrapper cursor = queryCrimes(CrimeTable.Cols.UUID + " = ?",
                new String[]{uuid.toString()})) {
            if (cursor.getCount() == 0) {
                return null;
            }
            cursor.moveToFirst();
            return cursor.getCrime();
        }
    }

    public void updateCrime(Crime c) {
        ContentValues values = getContentValues(c);
        mDatabase.update(CrimeTable.NAME, values, CrimeTable.Cols.UUID + " = ?",
                new String[]{c.getId().toString()});
    }

    public void addCrime(Crime c) {
        ContentValues values = getContentValues(c);
        mDatabase.insert(CrimeTable.NAME, null, values);
    }

    public void deleteCrime(Crime c) {
        mDatabase.delete(CrimeTable.NAME, CrimeTable.Cols.UUID + " = ?",
                new String[]{c.getId().toString()});
    }

    private static ContentValues getContentValues(Crime crime) {
        ContentValues values = new ContentValues();
        values.put(CrimeTable.Cols.UUID, crime.getId().toString());
        values.put(CrimeTable.Cols.TITLE, crime.getTitle());
        values.put(CrimeTable.Cols.DATE, crime.getDate().getTime());
        values.put(CrimeTable.Cols.SOLVED, crime.isSolved() ? 1 : 0);
        return values;
    }

    private CrimeCursorWrapper queryCrimes(String whereClause, String[] whereArgs) {
        Cursor cursor = mDatabase.query(CrimeTable.NAME,
                null,
                whereClause,
                whereArgs,
                null,
                null,
                null);
        return new CrimeCursorWrapper(cursor);
    }
}

File: app/src/main/java/com/work/criminalintent/kt/CrimeLab.kt
package com.work.criminalintent.kt

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import com.work.criminalintent.kt.database.CrimeBaseHelper
import com.work.criminalintent.kt.database.CrimeCursorWrapper
import com.work.criminalintent.kt.database.CrimeTable
import java.util.*
import kotlin.collections.ArrayList

class CrimeLab private constructor(context: Context) {

    private val mDatabase: SQLiteDatabase = CrimeBaseHelper(context).writableDatabase

    companion object {
        private var sCrimelab: CrimeLab? = null
        fun get(context: Context): CrimeLab = sCrimelab ?: CrimeLab(context)
    }

    fun getCrimes(): List<Crime> {
        val crimes = ArrayList<Crime>()
        queryCrimes(null, null).use { cursor ->
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                crimes.add(cursor.getCrime())
                cursor.moveToNext()
            }
        }
        return crimes
    }


    fun getCrime(uuid: UUID): Crime? {
        queryCrimes(CrimeTable.Cols.UUID + " = ?",
                arrayOf(uuid.toString())).use { cursor ->
            if (cursor.count == 0) return null
            cursor.moveToFirst()
            return cursor.getCrime()
        }
    }

    fun updateCrime(c: Crime) {
        val values = getContentValues(c)
        mDatabase.update(CrimeTable.NAME, values, CrimeTable.Cols.UUID + " = ?",
                arrayOf(c.id.toString()))
    }

    fun addCrime(c: Crime) {
        val values = getContentValues(c)
        mDatabase.insert(CrimeTable.NAME, null, values)
    }

    fun deleteCrime(c: Crime) =
            mDatabase.delete(CrimeTable.NAME, CrimeTable.Cols.UUID + " = ?",
                    arrayOf(c.id.toString()))


    private fun getContentValues(crime: Crime): ContentValues {
        return ContentValues().apply {
            put(CrimeTable.Cols.UUID, crime.id.toString())
            put(CrimeTable.Cols.TITLE, crime.title)
            put(CrimeTable.Cols.DATE, crime.date.time)
            put(CrimeTable.Cols.SOLVED, if (crime.isSolved) 1 else 0)
            put(CrimeTable.Cols.SUSPECT, crime.suspect)
        }
    }

    private fun queryCrimes(whereClause: String?, whereArgs: Array<String>?): CrimeCursorWrapper {
        val cursor: Cursor = mDatabase.query(CrimeTable.NAME,
                null,
                whereClause,
                whereArgs,
                null,
                null,
                null)
        return CrimeCursorWrapper(cursor)
    }
}